### ISSUE TITLE
Directory symlinks treated as file symlinks

### ISSUE THREAD

**Issue by michaldudak** [git-for-windows/git#1646]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.16.2.windows.1
cpu: x86_64
built from commit: e1848984d1004040ec5199e749b5f282ddf4bb09
sizeof-long: 4
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.17133.73]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VisualStudioCode
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I have Developer mode enabled in Windows

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
# creating the repo with files
mkdir repro
cd repro
git init
mkdir dir
echo hello > dir\file.txt
mklink file-link dir\file.txt
mklink /D dir-link dir
git add .
git commit -m "Initial commit"

# cloning the repo
cd ..
git clone -c core.symlinks=true .\repro cloned-repro
cd cloned-repro
dir
```
 - What did you expect to occur after running these commands?

`dir-link` should be a **\<SYMLINKD\>**

 - What actually happened instead?

`dir-link` is a **\<SYMLINK\>**. Because of this, Windows treats is as a link to a file (as if it was created with `mklink`, not `mklink /D`.

**Comment by orgads** [comment]
Possibly related: https://github.com/git-for-windows/git/issues/1197

**Comment by michaldudak** [comment]
Indeed. However, in my case, the directory already exists, so git should be able to determine the right type of symlink when creating it.

**Comment by orgads** [comment]
You cannot know if it exists, since this is a new clone. The symbolic link might get created before the directory...

**Comment by dscho** [comment]
> The symbolic link might get created before the directory...

We handle this by making a list of "phantom links" and processing them in an atexit() handler: https://github.com/git-for-windows/git/blob/df8884cbc5c39073848ddf2058bafeea1188312b/compat/mingw.c#L288-L326

@michaldudak any chance you can clone the Git for Windows SDK (via `git clone --depth 1 https://github.com/git-for-windows/git-sdk-64`), then build Git via `sdk build git` and then single-step through that function in GDB to figure out what is going wrong?

**Comment by michaldudak** [comment]
Sure, I'll try to do it in the evening.

**Comment by michaldudak** [comment]
Unfortunately, I haven't been able to debug it. I built Git as you described but GDB can't set any breakpoints (`Cannot insert breakpoint 1. Cannot access memory at address 0x58ec90`). I haven't used GDB before, so it' possible I'm doing something wrong, though.

**Comment by dscho** [comment]
> GDB can't set any breakpoints

Oh right... you need to remove the `-O2` from the `CFLAGS` in `/usr/src/git/Makefile` and then rebuild. Sorry, I keep forgetting about that.

FWIW I have this `config.mak` locally and compile with `make -j15 DEVELOPER=1 DEBUG=1` when I want to debug with GDB:

```
ASCIIDOC=asciidoctor
ASCIIDOC_HTML=html5
ASCIIDOC_DOCBOOK=docbook45
ASCIIDOC_EXTRA="-alitdd=&\#45;&\#45;"
ASCIIDOC_CONF=-I"/mingw64/lib/asciidoctor-extensions" -rman-inline-macro

DEVELOPER=1
ifdef DEBUG
CFLAGS := $(filter-out -O2,$(CFLAGS))
ASLR_OPTION := -Wl,--dynamicbase
BASIC_LDFLAGS := $(filter-out $(ASLR_OPTION),$(BASIC_LDFLAGS))
endif
```

**Comment by michaldudak** [comment]
Thanks, I'll try it again when I'm home in the evening as it behaves as expected on my work machine. It seems it may be something with Windows version (I'm using Insiders Preview at home).

**Comment by dscho** [comment]
Hmm. `2` means `ERROR_FILE_NOT_FOUND`. Reading lines `300-303` again, it strikes me as odd that we open `wlink` and not `wtarget`. The point is to determine whether the target of the link is a directory or a file, right?

What if you replace the `wlink` in the `CreateFileW()` call by `wtarget`?

**Comment by dscho** [comment]
(And if it fixes it, can you open a PR with a commit message that starts with the oneline `fixup! Win32: symlink: add support for symlinks to directories`?)

**Comment by michaldudak** [comment]
Indeed it works! Thanks for your help, I'll prepare a PR right away. I wonder why it always worked on my other machine, though...

**Comment by dscho** [comment]
> I wonder why it always worked on my other machine, though...

Maybe the link targets existed there already, for some reason? Have you tried on the other machine with a fresh clone?

**Comment by michaldudak** [comment]
Yes, it was a fresh clone. I'll investigate it in more depth when I have more time (possibly later today/tomorrow).

**Comment by dscho** [comment]
> I'll investigate it in more depth when I have more time

Thank you so much! If you don't mind, I'll hold off from merging your PR until we have clarity here (something feels off to me, and that is usually a good indicator that I do not understand the problem fully and therefore my suggested fix may very well be incomplete or wrong).

**Comment by michaldudak** [comment]
I've been able to isolate this issue. It seems that my newer Windows 10 (I don't have the possibility to check it on other machines at the moment, unfortunately) does not allow to call `CreateFile` on a link of a wrong type. I reproduced the issue by creating a file symlink to a directory (by `mklink` without any options passed) and trying to open it with CreateFileW with the same options as in Git. It works fine on Windows 1709 but fails on 1803 (with ERROR_ACCESS_DENIED). 
Your idea for a workaround seems to work fine, though. As Windows doesn't access the symlink, no error is thrown. If I have a chance to verify it on another updated machine, I'll do so.

**Comment by dscho** [comment]
@michaldudak great! Thank you!

Could I ask you two more favors? Can you add exactly that description why the bug was not caught earlier to the commit message, and then also add a Signed-off-by: line (that is what the DCO bot complained about)? Then I'll hit the merge button!

Thank you so much for working on this!

**Comment by dscho** [comment]
Oh, and a third thing: I would love it if you could have the line "This fixes https://github.com/git-for-windows/git/issues/1646" as last paragraph in the commit message ;-)

**Comment by michaldudak** [comment]
All done. Thanks for your assistance!
It was interesting to have something to do with C, after ~10 years of C# and JavaScript ;)

**Comment by dscho** [comment]
Thank you!

BTW if you're interested, I have a really challenging and rewarding project in Typescript in mind... something to do with contributing patches to git.git... Are you game?

**Comment by michaldudak** [comment]
Sounds interesting. I don't have massive amounts of free time recently, but I think I'd be able to help from time to time.

**Comment by michaldudak** [comment]
@dscho When do you plan to release GfW with this fix?

**Comment by adrach** [comment]
hi all, I just re-built and tried on Windows 10 April 2018 Update with Developer mode enabled, and I am still getting SYMLINK instead of SYMLINKD

**Comment by dscho** [comment]
@adrach could you provide a minimal example, please?

**Comment by adrach** [comment]
I used exactly the same minimal example as listed above

**Comment by dscho** [comment]
@adrach is this not the exact same example that now works for @michaldudak? Maybe you could spend a little more quality time with GDB to figure out what is going wrong for you?

**Comment by adrach** [comment]
yeah, I do not have tonight, but I will try to do it before the weekend

**Comment by dscho** [comment]
@adrach for the record, I just spent some quality time testing this. And in the interest of full transparency (which I recommend you follow in the future, too), here is *precisely* what I did.

First, I made sure that I had installed Git for Windows v2.17.0 by running `git version` in a CMD window, then ran

```cmd
mkdir repro
cd repro
git init
mkdir dir
echo hello > dir\file.txt
mklink file-link dir\file.txt
mklink /D dir-link dir
git add .
git commit -m "Initial commit"
cd ..
```

in a test directory and then ran the test:

```cmd
git clone repro repro-v2.17.0
dir repro-v2.17.0
```

Unsurprisingly, this resulted in the bug:

```
05/15/2018  08:34 AM    <DIR>          .
05/15/2018  08:34 AM    <DIR>          ..
05/15/2018  08:34 AM    <DIR>          dir
05/15/2018  08:34 AM    <SYMLINK>      dir-link [dir]
05/15/2018  08:34 AM    <SYMLINK>      file-link [dir\file.txt]
               2 File(s)              0 bytes
               3 Dir(s)  47,663,357,952 bytes free
```

Then, I built Git in my Git for Windows SDK from commit 05ca542f78d6eb85400bd10b0b60ce9442e5d76b and ran the following instructions in the CMD window to use the just-built Git:

```cmd
PATH=C:\git-sdk-64\usr\src\git;C:\git-sdk-64\mingw64\bin;C:\git-sdk-64\usr\bin;%PATH%
set GIT_EXEC_PATH=C:\git-sdk-64\usr\src\git
set MSYSTEM=MINGW
```

Again, I verified with `git version` that I was using the `git.exe` I intended, then ran

```cmd
git clone repro repro-with-fix
dir repro-with-fix
```

and now the bug was fixed! Verify for yourself.:

```
05/15/2018  08:38 AM    <DIR>          .
05/15/2018  08:38 AM    <DIR>          ..
05/15/2018  08:38 AM    <DIR>          dir
05/15/2018  08:38 AM    <SYMLINKD>     dir-link [dir]
05/15/2018  08:38 AM    <SYMLINK>      file-link [dir\file.txt]
               1 File(s)              0 bytes
               4 Dir(s)  47,048,884,224 bytes free
```

Now I reverted ec4523a1a7ac4883c93b434d543e2f61ddc9ef6e, rebuilt, again verified via `git version` that I was using the correctly-rebuilt `git.exe`, ran a similar clone and dir and it resulted in the bug again:

```
05/15/2018  08:40 AM    <DIR>          .
05/15/2018  08:40 AM    <DIR>          ..
05/15/2018  08:40 AM    <DIR>          dir
05/15/2018  08:40 AM    <SYMLINK>      dir-link [dir]
05/15/2018  08:40 AM    <SYMLINK>      file-link [dir\file.txt]
               2 File(s)              0 bytes
               3 Dir(s)  47,581,552,640 bytes free
```

Then, I installed the latest snapshot Portable Git as of https://wingit.blob.core.windows.net/files/PortableGit-prerelease-2.17.0.windows.1.38.g05ca542f78-64-bit.7z.exe into a sibling directory I called `portable-180503-64`, ran `portable-180503-64\cmd\git version` to verify that I use the Git version I intended, then ran

```cmd
portable-180503-64\cmd\git clone repro repro-with-snap
dir repro-with-snap
```

and it resulted in something completely different:

```
05/15/2018  08:46 AM    <DIR>          .
05/15/2018  08:46 AM    <DIR>          ..
05/15/2018  08:46 AM    <DIR>          dir
05/15/2018  08:46 AM                 3 dir-link
05/15/2018  08:46 AM                12 file-link
               2 File(s)             15 bytes
               3 Dir(s)  47,107,784,704 bytes free
```

This looked like the symbolic links were not created at all, so I called `portable-180503-64\cmd\git config --show-origin core.symlinks` and sure enough, the Portable Git has it hard-coded to `false`. I forced it to `true` with `portable-180503-64\cmd\git config --system core.symlinks true`, then ran

```cmd
portable-180503-64\cmd\git clone repro repro-with-snap2
dir repro-with-snap2
```

and was rewarded with the expected output (*including* `SYMLINKD`):

```
05/15/2018  08:48 AM    <DIR>          .
05/15/2018  08:48 AM    <DIR>          ..
05/15/2018  08:48 AM    <DIR>          dir
05/15/2018  08:48 AM    <SYMLINKD>     dir-link [dir]
05/15/2018  08:48 AM    <SYMLINK>      file-link [dir\file.txt]
               1 File(s)              0 bytes
               4 Dir(s)  47,107,739,648 bytes free
```

My `ver` says that this is `Microsoft Windows [Version 10.0.17134.48]`, just in case that makes a difference (which I do *not* expect).

In short: my findings contradict your findings starkly, @adrach.

**Comment by dscho** [comment]
>> I have a really challenging and rewarding project in Typescript in mind... something to do with contributing patches to git.git... Are you game?
>
> Sounds interesting. I don't have massive amounts of free time recently, but I think I'd be able to help from time to time.

Okay, could you have a look at https://github.com/gitgitgadget/gitgitgadget and tell me everything that's wrong with it, preferably in a new issue? (I'll be mostly offline for a week, but will be eager to read what you have to say about the gadget afterwards.)

**Comment by michaldudak** [comment]
Hi, unfortunately June is pretty busy for me and I won't be able to work on anything extra at the moment. July seems more probable.
